File: tests/unit/installer/options.test.ts
import {
	afterEach,
	beforeEach,
	describe,
	expect,
	mock,
	spyOn,
	test,
} from "bun:test";
import process from "node:process";
import {
	dockerFlavourOf,
	OptionsParser,
} from "../../../packages/installer/options";

describe("OptionsParser.parseArgs", () => {
	test("defaults when no flags are given", () => {
		expect(OptionsParser.parseArgs([])).toEqual({
			agentPort: 7420,
			dryRun: false,
			migrateToRootful: false,
			mode: "agent",
			rootlessUser: "homerun",
			version: "latest",
			yes: false,
		});
	});

	test("--dry-run sets dryRun", () => {
		expect(OptionsParser.parseArgs(["--dry-run"]).dryRun).toBe(true);
	});

	test("--yes and -y both set yes", () => {
		expect(OptionsParser.parseArgs(["--yes"]).yes).toBe(true);
		expect(OptionsParser.parseArgs(["-y"]).yes).toBe(true);
	});

	test("--mode=full and --mode=agent set mode", () => {
		expect(OptionsParser.parseArgs(["--mode=full"]).mode).toBe("full");
		expect(OptionsParser.parseArgs(["--mode=agent"]).mode).toBe("agent");
	});

	test("--docker=rootful and --docker=rootless set docker", () => {
		expect(OptionsParser.parseArgs(["--docker=rootful"]).docker).toBe(
			"rootful",
		);
		expect(
			OptionsParser.parseArgs(["--docker=rootful", "--docker=rootless"]).docker,
		).toBe("rootless");
	});

	test("--user=<name> overrides rootlessUser", () => {
		expect(OptionsParser.parseArgs(["--user=myuser"]).rootlessUser).toBe(
			"myuser",
		);
	});

	test("--version=<tag> overrides version", () => {
		expect(OptionsParser.parseArgs(["--version=v1.2.3"]).version).toBe(
			"v1.2.3",
		);
	});

	test("--port=<n> overrides agentPort", () => {
		expect(OptionsParser.parseArgs(["--port=9000"]).agentPort).toBe(9000);
	});

	test("combines multiple flags", () => {
		const opts = OptionsParser.parseArgs([
			"--dry-run",
			"--mode=full",
			"--user=alice",
			"--version=v2.0.0",
			"--port=8080",
			"--yes",
		]);
		expect(opts).toEqual({
			agentPort: 8080,
			dryRun: true,
			migrateToRootful: false,
			mode: "full",
			rootlessUser: "alice",
			version: "v2.0.0",
			yes: true,
		});
	});
});

describe("OptionsParser.parseArgs exit paths", () => {
	let exitSpy: ReturnType<typeof spyOn>;
	let logSpy: ReturnType<typeof spyOn>;
	let errorSpy: ReturnType<typeof spyOn>;

	beforeEach(() => {
		exitSpy = spyOn(process, "exit").mockImplementation(
			(() => undefined) as never,
		);
		logSpy = spyOn(console, "log").mockImplementation(() => undefined);
		errorSpy = spyOn(console, "error").mockImplementation(() => undefined);
	});

	afterEach(() => {
		mock.restore();
	});

	test("--help prints help and exits 0", () => {
		OptionsParser.parseArgs(["--help"]);
		expect(logSpy).toHaveBeenCalled();
		expect(logSpy.mock.calls[0][0]).toContain("homerun-install");
		expect(exitSpy).toHaveBeenCalledWith(0);
	});

	test("-h behaves the same as --help", () => {
		OptionsParser.parseArgs(["-h"]);
		expect(exitSpy).toHaveBeenCalledWith(0);
	});

	test("an unknown argument prints an error and exits 1", () => {
		OptionsParser.parseArgs(["--bogus"]);
		expect(errorSpy).toHaveBeenCalled();
		expect(errorSpy.mock.calls[0][0]).toContain("Unknown argument: --bogus");
		expect(exitSpy).toHaveBeenCalledWith(1);
	});
});

describe("OptionsParser.printHelp", () => {
	test("mentions every documented flag", () => {
		const logSpy = spyOn(console, "log").mockImplementation(() => undefined);
		OptionsParser.printHelp();
		const output = logSpy.mock.calls.map((c) => c[0]).join("\n");
		logSpy.mockRestore();

		for (const flag of [
			"--version=",
			"--mode=agent|full",
			"--docker=rootless|rootful",
			"--advertise-addr=",
			"--migrate-to-rootful",
			"--image=",
			"--user=",
			"--port=",
			"--dry-run",
			"--yes",
		]) {
			expect(output).toContain(flag);
		}
	});
});

describe("OptionsParser.parseArgs --domain", () => {
	test("is unset by default and takes the value verbatim", () => {
		expect(OptionsParser.parseArgs([]).domain).toBeUndefined();
		expect(
			OptionsParser.parseArgs(["--domain=https://homerun.example.com/"]).domain,
		).toBe("https://homerun.example.com/");
	});
});

describe("OptionsParser.parseArgs swarm and migration flags", () => {
	test("--advertise-addr and --image take their values", () => {
		const opts = OptionsParser.parseArgs([
			"--advertise-addr=10.0.0.5",
			"--image=homerun:local",
		]);
		expect(opts.advertiseAddress).toBe("10.0.0.5");
		expect(opts.image).toBe("homerun:local");
	});

	test("--migrate-to-rootful implies a full install", () => {
		const opts = OptionsParser.parseArgs(["--migrate-to-rootful"]);
		expect(opts.migrateToRootful).toBe(true);
		expect(opts.mode).toBe("full");
	});
});

describe("dockerFlavourOf", () => {
	test("a full install defaults to the system daemon", () => {
		expect(dockerFlavourOf(OptionsParser.parseArgs(["--mode=full"]))).toBe(
			"rootful",
		);
	});

	test("a full install can opt into rootless", () => {
		expect(
			dockerFlavourOf(
				OptionsParser.parseArgs(["--mode=full", "--docker=rootless"]),
			),
		).toBe("rootless");
	});

	test("the agent always runs rootless", () => {
		expect(dockerFlavourOf(OptionsParser.parseArgs([]))).toBe("rootless");
	});
});

describe("OptionsParser.validate", () => {
	test("accepts rootful Docker for a full install", () => {
		expect(
			OptionsParser.validate(
				OptionsParser.parseArgs(["--mode=full", "--docker=rootful"]),
			),
		).toBeNull();
	});

	test("rejects an explicit daemon for an agent install", () => {
		expect(
			OptionsParser.validate(OptionsParser.parseArgs(["--docker=rootful"])),
		).toContain("--docker=rootful only applies to --mode=full");
		expect(
			OptionsParser.validate(OptionsParser.parseArgs(["--docker=rootless"])),
		).toContain("--docker=rootless only applies to --mode=full");
	});

	test("rejects migrating to rootful while asking for rootless", () => {
		expect(
			OptionsParser.validate(
				OptionsParser.parseArgs(["--migrate-to-rootful", "--docker=rootless"]),
			),
		).toContain("drop --docker=rootless");
	});

	test("rejects --advertise-addr without a swarm to advertise", () => {
		expect(
			OptionsParser.validate(
				OptionsParser.parseArgs([
					"--mode=full",
					"--docker=rootless",
					"--advertise-addr=10.0.0.5",
				]),
			),
		).toContain("--advertise-addr only applies");
		expect(
			OptionsParser.validate(
				OptionsParser.parseArgs(["--mode=full", "--advertise-addr=10.0.0.5"]),
			),
		).toBeNull();
	});

	test("accepts the defaults and a plain migration", () => {
		expect(OptionsParser.validate(OptionsParser.parseArgs([]))).toBeNull();
		expect(
			OptionsParser.validate(OptionsParser.parseArgs(["--migrate-to-rootful"])),
		).toBeNull();
	});
});

File: cmd/installer/options_test.go
package main

import (
	"strings"
	"testing"
)

func TestParseArgsDefaults(t *testing.T) {
	opts, help, err := ParseArgs(nil)
	if err != nil || help {
		t.Fatalf("no arguments should parse cleanly, got help=%t err=%v", help, err)
	}
	if opts.Mode != ModeAgent {
		t.Errorf("default mode should be agent, got %q", opts.Mode)
	}
	if opts.AgentPort != 7420 {
		t.Errorf("default port should be 7420, got %d", opts.AgentPort)
	}
	if opts.RootlessUser != "homerun" {
		t.Errorf("default user should be homerun, got %q", opts.RootlessUser)
	}
	if opts.Version != "latest" {
		t.Errorf("default version should be latest, got %q", opts.Version)
	}
	if opts.DryRun || opts.Yes || opts.MigrateToRootful {
		t.Errorf("no boolean flag should default on, got %+v", opts)
	}
}

func TestParseArgsFlags(t *testing.T) {
	opts, _, err := ParseArgs([]string{
		"--mode=full", "--docker=rootless", "--dry-run", "-y",
		"--domain=homerun.example.com", "--user=deployer", "--port=9000",
		"--version=v1.2.3", "--image=local/homerun:dev", "--advertise-addr=10.0.0.5",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	for name, check := range map[string]bool{
		"mode":      opts.Mode == ModeFull,
		"docker":    opts.Docker == FlavourRootless,
		"dry run":   opts.DryRun,
		"yes":       opts.Yes,
		"domain":    opts.Domain == "homerun.example.com",
		"user":      opts.RootlessUser == "deployer",
		"port":      opts.AgentPort == 9000,
		"version":   opts.Version == "v1.2.3",
		"image":     opts.Image == "local/homerun:dev",
		"advertise": opts.AdvertiseAddress == "10.0.0.5",
	} {
		if !check {
			t.Errorf("%s not parsed: %+v", name, opts)
		}
	}
}

func TestParseArgsMigrateImpliesFullMode(t *testing.T) {
	opts, _, err := ParseArgs([]string{"--migrate-to-rootful"})
	if err != nil {
		t.Fatal(err)
	}
	if !opts.MigrateToRootful || opts.Mode != ModeFull {
		t.Errorf("--migrate-to-rootful implies --mode=full, got %+v", opts)
	}
}

func TestParseArgsHelpAndErrors(t *testing.T) {
	for _, arg := range []string{"--help", "-h"} {
		if _, help, err := ParseArgs([]string{arg}); !help || err != nil {
			t.Errorf("%s should ask for help, got help=%t err=%v", arg, help, err)
		}
	}

	if _, _, err := ParseArgs([]string{"--nope"}); err == nil {
		t.Error("an unknown argument should be rejected")
	} else if !strings.Contains(err.Error(), "--nope") {
		t.Errorf("the error should name the argument, got %q", err)
	}

	if _, _, err := ParseArgs([]string{"--port=many"}); err == nil {
		t.Error("a non-numeric --port= should be rejected")
	}
}

func TestDockerFlavourOf(t *testing.T) {
	cases := []struct {
		name string
		opts Options
		want DockerFlavour
	}{
		{"agent is always rootless", Options{Mode: ModeAgent}, FlavourRootless},
		{"agent ignores an explicit flavour", Options{Mode: ModeAgent, Docker: FlavourRootful}, FlavourRootless},
		{"full defaults to rootful", Options{Mode: ModeFull}, FlavourRootful},
		{"full honours rootless", Options{Mode: ModeFull, Docker: FlavourRootless}, FlavourRootless},
	}
	for _, testCase := range cases {
		if got := DockerFlavourOf(testCase.opts); got != testCase.want {
			t.Errorf("%s: want %q, got %q", testCase.name, testCase.want, got)
		}
	}
}

func TestValidate(t *testing.T) {
	cases := []struct {
		name    string
		opts    Options
		wantErr string
	}{
		{"valid agent install", Options{Mode: ModeAgent}, ""},
		{"valid full install", Options{Mode: ModeFull, Docker: FlavourRootful}, ""},
		{
			"--docker on an agent install",
			Options{Mode: ModeAgent, Docker: FlavourRootful},
			"only applies to --mode=full",
		},
		{
			"--migrate-to-rootful with --docker=rootless",
			Options{Mode: ModeFull, Docker: FlavourRootless, MigrateToRootful: true},
			"drop --docker=rootless",
		},
		{
			"--advertise-addr without a swarm",
			Options{Mode: ModeFull, Docker: FlavourRootless, AdvertiseAddress: "10.0.0.5"},
			"only applies to a rootful --mode=full install",
		},
		{
			"--advertise-addr on an agent install",
			Options{Mode: ModeAgent, AdvertiseAddress: "10.0.0.5"},
			"only applies to a rootful --mode=full install",
		},
	}
	for _, testCase := range cases {
		got := Validate(testCase.opts)
		if testCase.wantErr == "" && got != "" {
			t.Errorf("%s: expected valid, got %q", testCase.name, got)
		}
		if testCase.wantErr != "" && !strings.Contains(got, testCase.wantErr) {
			t.Errorf("%s: want a message containing %q, got %q", testCase.name, testCase.wantErr, got)
		}
	}
}

func TestHelpTextCoversEveryFlag(t *testing.T) {
	for _, flag := range []string{
		"--version=", "--mode=agent|full", "--domain=", "--docker=rootless|rootful",
		"--advertise-addr=", "--migrate-to-rootful", "--image=", "--user=", "--port=",
		"--dry-run", "--yes",
	} {
		if !strings.Contains(helpText, flag) {
			t.Errorf("--help doesn't document %s", flag)
		}
	}
}
